### PR TITLE
Detect if the node's kubelet MountPropagation option is enabled before deploy CSI

### PR DESCRIPTION
**Pull request by niusmallnan**
~1. use the configz to detect if the user has enabled MountPropagation~
~2. create a pod and get back pod spec and check if all the MountPropagation was set correctly~
~3. add nodes/proxy for accessing node configz~

Sometimes the kubelet configuration on different nodes are different, so we need to create a daemonset and check the pod spec.

https://github.com/rancher/longhorn/issues/108

### CODE REVIEW

File: app/driver.go [hunk 1]
```diff
@@ -37,10 +37,12 @@ const (
 	FlagCSIProvisionerImage     = "csi-provisioner-image"
 	FlagCSIDriverRegistrarImage = "csi-driver-registrar-image"
 	FlagCSIProvisionerName      = "csi-provisioner-name"
+	FlagCSIPodCheckImage        = "csi-pod-check-image"
```
Comment (yasker) [inline app/driver.go]: Just reuse the manager image. It's already on the node and needs to be run anyway.

File: csi/deployment_util.go [hunk 2]
```diff
@@ -10,8 +10,79 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/pointer"
 )
 
+// CheckMountPropagationWithPodSpec https://github.com/kubernetes/kubernetes/issues/66086#issuecomment-404346854
+func CheckMountPropagationWithPodSpec(kubeClient *clientset.Clientset, image, namespace string) (bool, error) {
+	commonName := "pod-mount-test"
```
Comment (yasker) [inline csi/deployment_util.go]: We can use the name `longhorn-mount-propagation-tester`

File: csi/deployment_util.go [hunk 3]
```diff
@@ -10,8 +10,79 @@ import (
...
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				v1.Container{
+					Name:  commonName,
+					Image: image,
+					Args: []string{
+						"/bin/sh",
+						"-c",
+						"sleep 1000000000",
```
Comment (yasker) [inline csi/deployment_util.go]: `sleep` in longhorn-manager image can do `sleep infinity`

**Comment by niusmallnan** [comment]
Created a GKE cluster with six nodes:
```
# kubectl get no
NAME                                        STATUS    ROLES     AGE       VERSION
gke-zhibo-test-default-pool-b402ba57-12jw   Ready     <none>    20m       v1.10.5-gke.0
gke-zhibo-test-default-pool-b402ba57-3qwp   Ready     <none>    20m       v1.10.5-gke.0
gke-zhibo-test-default-pool-b402ba57-8lv6   Ready     <none>    20m       v1.10.5-gke.0
gke-zhibo-test-default-pool-b402ba57-9dvv   Ready     <none>    20m       v1.10.5-gke.0
gke-zhibo-test-default-pool-b402ba57-k6tn   Ready     <none>    20m       v1.10.5-gke.0
gke-zhibo-test-default-pool-b402ba57-nl82   Ready     <none>    20m       v1.10.5-gke.0

```


Can get these logs from driver-deployer:
```
# kubectl logs po/longhorn-driver-deployer-58754748d-sfglg -n longhorn-system
DEBU[2018-07-18T06:46:50Z] Deploying CSI driver
DEBU[2018-07-18T06:46:50Z] Trying to create a daemonset longhorn-mount-propagation-tester for MountPropagation checking
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-5j6wn, node gke-zhibo-test-default-pool-b402ba57-k6tn
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-945rs, node gke-zhibo-test-default-pool-b402ba57-12jw
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-dbbwl, node gke-zhibo-test-default-pool-b402ba57-8lv6
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-fw8fl, node gke-zhibo-test-default-pool-b402ba57-9dvv
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-jj5kg, node gke-zhibo-test-default-pool-b402ba57-nl82
DEBU[2018-07-18T06:46:50Z] Got MountPropagation Bidirectional from pod longhorn-mount-propagation-tester-mjfnb, node gke-zhibo-test-default-pool-b402ba57-3qwp
DEBU[2018-07-18T06:46:50Z] Trying to get service csi-attacher
DEBU[2018-07-18T06:46:50Z] Trying to create service csi-attacher
DEBU[2018-07-18T06:46:50Z] Trying to get statefulSet csi-attacher
DEBU[2018-07-18T06:46:50Z] Trying to create statefulSet csi-attacher
DEBU[2018-07-18T06:46:50Z] Trying to get service csi-provisioner
DEBU[2018-07-18T06:46:50Z] Trying to create service csi-provisioner
DEBU[2018-07-18T06:46:50Z] Trying to get statefulSet csi-provisioner
DEBU[2018-07-18T06:46:50Z] Trying to create statefulSet csi-provisioner
DEBU[2018-07-18T06:46:50Z] Trying to get daemonset longhorn-csi-plugin
DEBU[2018-07-18T06:46:50Z] Trying to create daemonset longhorn-csi-plugin
```

File: app/driver.go [hunk 4]
```diff
@@ -148,11 +148,16 @@ func chooseDriver(kubeClient *clientset.Clientset) (string, error) {
 }
 
 func deployCSIDriver(kubeClient *clientset.Clientset, c *cli.Context, managerImage string) error {
+	namespace := os.Getenv(types.EnvPodNamespace)
+
+	if checked, err := csi.CheckMountPropagationWithPodSpec(kubeClient, managerImage, namespace); err != nil || !checked {
```
Comment (yasker) [inline app/driver.go]: I thought it should be a switch in the chooseDriver. If mount propagation check failed, we should fall back to Flexvolume (and explicitly tell user in the log what's happened)

Comment (yasker) [reply app/driver.go]: Also we need to consider if user explicitly choose CSI but we cannot deploy. In this case we should error out instead.

**Comment by niusmallnan** [comment]
Case: set MountPropagation to false and let driver-deployer detect the driver 
```
# kubectl logs  pod/longhorn-driver-deployer-866f9d5b-dv4pz -n longhorn-system
DEBU[2018-07-18T12:34:52Z] Driver csi will be used after automatic detection
DEBU[2018-07-18T12:34:52Z] Trying to create the daemonset longhorn-mount-propagation-tester for MountPropagation check
DEBU[2018-07-18T12:34:52Z] Created the daemonset longhorn-mount-propagation-tester for MountPropagation check
DEBU[2018-07-18T12:34:52Z] Got MountPropagation <nil> from pod longhorn-mount-propagation-tester-cwfmp, node 172.31.27.162
INFO[2018-07-18T12:34:52Z] MountPropagation check failed, should fall back to the flexvolume
DEBU[2018-07-18T12:34:52Z] Deploying Flexvolume driver
INFO[2018-07-18T12:34:52Z] User specified Flexvolume dir at: /var/lib/kubelet/volumeplugins
INFO[2018-07-18T12:34:52Z] Install Flexvolume to Kubernetes nodes directory /var/lib/kubelet/volumeplugins
```

Case: set MountPropagation to false and explicitly choose CSI
```
# kubectl logs pod/longhorn-driver-deployer-dd54948d9-7jwcb -n longhorn-system
DEBU[2018-07-18T12:49:22Z] Explicitly choose the csi driver
DEBU[2018-07-18T12:49:22Z] Trying to create the daemonset longhorn-mount-propagation-tester for MountPropagation check
DEBU[2018-07-18T12:49:22Z] Created the daemonset longhorn-mount-propagation-tester for MountPropagation check
DEBU[2018-07-18T12:49:22Z] Got MountPropagation <nil> from pod longhorn-mount-propagation-tester-6sjgm, node 172.31.27.162
FATA[2018-07-18T12:49:22Z] Error deploying driver: CSI cannot be deployed because MountPropagation is not set on kubelet and api-server
```

File: app/driver.go [hunk 5]
```diff
@@ -109,12 +109,35 @@ func deployDriver(c *cli.Context) error {
...
 	driver := c.String(FlagDriver)
 	if driver == "" {
+		driverDetected = true
 		driver, err = chooseDriver(kubeClient)
+		logrus.Debugf("Driver %s will be used after automatic detection", driver)
 		if err != nil {
 			return err
 		}
+	} else {
+		logrus.Debugf("Explicitly choose the %s driver", driver)
```
Comment (yasker) [inline app/driver.go]: `User specified the driver %s`

File: app/driver.go [hunk 6]
```diff
@@ -109,12 +109,35 @@ func deployDriver(c *cli.Context) error {
...
+	}
+
+	if driver == FlagDriverCSI {
+		csiChecked, err := csi.CheckMountPropagationWithPodSpec(kubeClient, managerImage, os.Getenv(types.EnvPodNamespace))
```
Comment (yasker) [inline app/driver.go]: The function can just return an error, including when it's detected mount propagation was not enabled, since the case that `false, nil` is handled in the same way as `nil, err`.

File: app/driver.go [hunk 7]
```diff
@@ -109,12 +109,35 @@ func deployDriver(c *cli.Context) error {
...
+		}
+
+		// if user explicitly choose CSI but we cannot deploy.
+		// In this case we should error out instead.
+		if !csiChecked {
+			if driverDetected {
+				logrus.Infof("MountPropagation check failed, should fall back to the flexvolume")
+				driver = FlagDriverFlexvolume
+			} else {
+				return errors.New("CSI cannot be deployed because MountPropagation is not set on kubelet and api-server")
```
Comment (yasker) [inline app/driver.go]: Use `fmt.Errorf` instead. It's an item checked by go report (https://goreportcard.com/report/github.com/rancher/longhorn-manager).

File: app/driver.go [hunk 8]
```diff
@@ -109,12 +109,35 @@ func deployDriver(c *cli.Context) error {
...
+		csiChecked, err := csi.CheckMountPropagationWithPodSpec(kubeClient, managerImage, os.Getenv(types.EnvPodNamespace))
+		if err != nil {
+			logrus.Warnf("Got an error when checking MountPropagation with pod spec, %v", err)
+		}
+
+		// if user explicitly choose CSI but we cannot deploy.
+		// In this case we should error out instead.
+		if !csiChecked {
+			if driverDetected {
+				logrus.Infof("MountPropagation check failed, should fall back to the flexvolume")
```
Comment (yasker) [inline app/driver.go]: `MountPropagation check failed, fall back to use the Flexvolume`

File: csi/deployment_util.go [hunk 9]
```diff
@@ -10,8 +10,110 @@ import (
...
+
+	defer func() {
+		propagation := metav1.DeletePropagationForeground
+		if err := daemonSetClient.Delete(commonName,
+			&metav1.DeleteOptions{PropagationPolicy: &propagation}); err != nil {
+			logrus.Warnf("Failed to delete the daemonset %s for MountPropagation check", commonName)
+		}
+	}()
+
+	pods, err := podClient.List(metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", commonName)})
```
Comment (yasker) [inline csi/deployment_util.go]: It may not be shown immediately in the pod list like the issue Jacie met before. Retry for a short period of time (e.g. 30s) if the pod doesn't show up.

File: csi/deployment_util.go [hunk 10]
```diff
@@ -89,22 +191,24 @@ func getCommondStatefulSet(commonName, namespace, serviceAccount, image string,
 }
 
 func cleanupService(kubeClient *clientset.Clientset, service *v1.Service) error {
-	logrus.Debugf("Trying to get service %s", service.ObjectMeta.Name)
+	logrus.Debugf("Trying to get the service %s", service.ObjectMeta.Name)
```
Comment (yasker) [inline csi/deployment_util.go]: Separate the log changes into another commit.

File: csi/deployment_util.go (hunk 9 above)
Comment (niusmallnan) [reply csi/deployment_util.go]: I have considered this problem, but cannot reproduce it on GKE.
It's ok, I added the retry logic.

File: app/driver.go [hunk 11]
```diff
@@ -109,12 +109,32 @@ func deployDriver(c *cli.Context) error {
...
 	driver := c.String(FlagDriver)
 	if driver == "" {
+		driverDetected = true
 		driver, err = chooseDriver(kubeClient)
+		logrus.Debugf("Driver %s will be used after automatic detection", driver)
 		if err != nil {
 			return err
 		}
+	} else {
+		logrus.Debugf("User specified the drive %s", driver)
```
Comment (yasker) [inline app/driver.go]: Typo. `driver`

File: csi/deployment_util.go [hunk 12]
```diff
@@ -10,8 +11,142 @@ import (
...
+
+	retryCount := 0
+	for {
+		if retryCount >= maxRetryCountForMountPropagationCheck {
+			return fmt.Errorf("Has been retried %d times, but daemonset is still unavailable", retryCount)
+		}
+		time.Sleep(durationSleepForMountPropagationCheck)
+		retryCount++
+		ds, err := daemonSetClient.Get(commonName, metav1.GetOptions{})
+		if err != nil {
```
Comment (yasker) [inline csi/deployment_util.go]: I think the only valid error here is `apierrors.IsNotFound(err)`. Otherwise, it would be a real problem and we probably don't want to retry.

Comment (niusmallnan) [reply csi/deployment_util.go]: We are getting the daemonset not its pods, is it possible to get IsNotFound error?

Comment (yasker) [reply csi/deployment_util.go]: The race condition is due to a gap between the API server receive the request to create the object, and it actually creates the object and makes the client known. The race condition Jacie saw last time is for replica CRD object, so I think it should applies to all the object. Seems creation request is not a blocking command and won't guarantee the immediate request after creation can see the object ready.

Comment (niusmallnan) [reply csi/deployment_util.go]: Done.

But this may be slightly different, Jacie used List, and here is Get.
List reads data from the cache, so sometimes it is not possible to get the data right away.
But if adding this code to every Get, it looks ugly.

I will confirm how to use Get properly.

Comment (yasker) [reply csi/deployment_util.go]: Yeah, that's possible too. But even for the lister, it shouldn't so hard to use as well. I thought maybe it's because of the eventual consistency of the etcd, but not entirely sure about that.

Please let me know what you find in the end.

**Comment by yasker** [comment]
Merged, thanks.
